iocore/net/quic/QUICLossDetector.cc

```diff
@@ -500,50 +456,33 @@ QUICLossDetector::_detect_lost_packets(QUICPacketNumber largest_acked_packet_num
 // ===== Functions below are used on the spec but there're no pseudo code  =====
 
 void
-QUICLossDetector::_retransmit_all_unacked_handshake_data()
+QUICLossDetector::_retransmit_all_unacked_crypto_data()
 {
   SCOPED_MUTEX_LOCK(transmitter_lock, this->_transmitter->get_packet_transmitter_mutex().get(), this_ethread());
   SCOPED_MUTEX_LOCK(lock, this->_loss_detection_mutex, this_ethread());
-  std::set<QUICPacketNumber> retransmitted_handshake_packets;
+  std::set<QUICPacketNumber> retransmitted_crypto_packets;
   std::map<QUICPacketNumber, PacketInfo *> lost_packets;
 
   for (auto &info : this->_sent_packets) {
-    retransmitted_handshake_packets.insert(info.first);
-    lost_packets.insert({info.first, info.second.get()});
-    this->_transmitter->retransmit_packet(*info.second->packet);
+    if (info.second->is_crypto_packet) {
+      retransmitted_crypto_packets.insert(info.first);
+      this->_retransmit_lost_packet(info.second->packet);
+      lost_packets.insert({info.first, info.second.get()});
+    }
   }
 
-  this->_cc->on_packets_lost(lost_packets);
-  for (auto packet_number : retransmitted_handshake_packets) {
+  this->_cc->on_packets_lost(lost_packets, this->_pto_count);
+  for (auto packet_number : retransmitted_crypto_packets) {
     this->_remove_from_sent_packet_list(packet_number);
   }
 }
 
-void
-QUICLossDetector::_send_one_packet()
-{
-  SCOPED_MUTEX_LOCK(transmitter_lock, this->_transmitter->get_packet_transmitter_mutex().get(), this_ethread());
-  SCOPED_MUTEX_LOCK(lock, this->_loss_detection_mutex, this_ethread());
-
-  auto ite = this->_sent_packets.rbegin();
-  if (ite != this->_sent_packets.rend()) {
-    this->_transmitter->retransmit_packet(*ite->second->packet);
-  }
-}
-
 void
 QUICLossDetector::_send_two_packets()
 {
   SCOPED_MUTEX_LOCK(transmitter_lock, this->_transmitter->get_packet_transmitter_mutex().get(), this_ethread());
   SCOPED_MUTEX_LOCK(lock, this->_loss_detection_mutex, this_ethread());
-  auto ite = this->_sent_packets.rbegin();
-  if (ite != this->_sent_packets.rend()) {
-    this->_transmitter->retransmit_packet(*ite->second->packet);
-    ite++;
-    if (ite != this->_sent_packets.rend()) {
-      this->_transmitter->retransmit_packet(*ite->second->packet);
-    }
-  }
+  // TODO sent ping
```

Comment:
We need to sent two ack_eliciting packets. Not sure here, Ping frame might be a good idea.

Reply:
Maybe. It's a tricky requirement. Let's leave it for now. Can you create an issue for this?